### ISSUE TITLE
Allow sys.stdout to be replaced

### ISSUE THREAD

**Issue by alanjds** [grumpyhome/grumpy#88]
Please refer to [the discussion](https://github.com/google/grumpy/issues/290#issuecomment-297933648) on google/grumpy#290

Do not merge until `__getattr__` and `__setattr__` is working.

**Comment by alanjds** [comment]
<a href="https://github.com/corona10"><img src="https://avatars2.githubusercontent.com/u/5110323?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [corona10](https://github.com/corona10)**
_Friday Apr 28, 2017 at 15:51 GMT_

----

I have one thing to discuss here. 
At this point, I think that we should create `module` directory for these kinds of task. `runtime/sysmodule.go` => `module/sysmodule.go`
WDYT? @trotterdylan @alanjds

**Comment by alanjds** [comment]
<a href="https://github.com/alanjds"><img src="https://avatars3.githubusercontent.com/u/155623?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [alanjds](https://github.com/alanjds)**
_Friday Apr 28, 2017 at 16:25 GMT_

----

I totally agree, but am not mature enough on Golang to glance the implications of this change.

**Comment by alanjds** [comment]
<a href="https://github.com/corona10"><img src="https://avatars2.githubusercontent.com/u/5110323?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [corona10](https://github.com/corona10)**
_Friday Apr 28, 2017 at 16:27 GMT_

----

I believe that more modules will be needed. e.g (socketModule etc..)

**Comment by alanjds** [comment]
<a href="https://github.com/alanjds"><img src="https://avatars3.githubusercontent.com/u/155623?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [alanjds](https://github.com/alanjds)**
_Saturday Apr 29, 2017 at 20:55 GMT_

----

This is no more a WIP, after changed from `__getattr__` to `__getattribute__` fixed it.

Will continue the google/grumpy#290 implementation over this code.

Please discuss what changes we want before this (#70) got be merged.

**Comment by alanjds** [comment]
<a href="https://github.com/trotterdylan"><img src="https://avatars2.githubusercontent.com/u/117627?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [trotterdylan](https://github.com/trotterdylan)**
_Saturday Apr 29, 2017 at 23:28 GMT_

----

I would avoid \_\_getattribute\_\_ and instead use \_\_getattr\_\_. The latter is called only after nothing is found through normal attribute lookup, which is probably what we want in this case.

**Comment by alanjds** [comment]
<a href="https://github.com/alanjds"><img src="https://avatars3.githubusercontent.com/u/155623?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [alanjds](https://github.com/alanjds)**
_Sunday Apr 30, 2017 at 01:48 GMT_

----

Me too. But this rabbit hole is getting too deep!
Let me close some Print() issues first and later I come back to implement `__getattr__`.

This `__getattr__` issue seems at least one magnitude harder than the other issues anyway. I will need to get "slots" understanding first.

**Comment by alanjds** [comment]
<a href="https://github.com/alanjds"><img src="https://avatars3.githubusercontent.com/u/155623?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [alanjds](https://github.com/alanjds)**
_Monday May 01, 2017 at 19:00 GMT_

----

Just to say, tracebacks would be very nice. Took me quite a bit to realize this 'res' typo

**Comment by alanjds** [comment]
<a href="https://github.com/alanjds"><img src="https://avatars3.githubusercontent.com/u/155623?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [alanjds](https://github.com/alanjds)**
_Saturday Jun 10, 2017 at 15:48 GMT_

----

> I would avoid __getattribute__ and instead use __getattr__. The latter is called only after nothing is found through normal attribute lookup, which is probably what we want in this case.

@trotterdylan Are you ok with going on this way, and allow to use __getattribute__ for now, or is better to go develop __getattr__ before keep going on this PR?

**Comment by alanjds** [comment]
<a href="https://github.com/trotterdylan"><img src="https://avatars2.githubusercontent.com/u/117627?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [trotterdylan](https://github.com/trotterdylan)**
_Wednesday Jun 14, 2017 at 14:35 GMT_

----

I think that \_\_getattr\_\_ should be fairly straightforward to implement. It needs a new slot and fallback logic in grumpy.GetAttr(). Also a few tests that validate the correct attribute resolution.

**Comment by alanjds** [comment]
<a href="https://github.com/alanjds"><img src="https://avatars3.githubusercontent.com/u/155623?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [alanjds](https://github.com/alanjds)**
_Friday Jun 16, 2017 at 16:40 GMT_

----

Ok. Then I will chase \_\_getattr__ on another PR.

Everything else is ok?

**Comment by alanjds** [comment]
<a href="https://github.com/trotterdylan"><img src="https://avatars2.githubusercontent.com/u/117627?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [trotterdylan](https://github.com/trotterdylan)**
_Monday Jun 19, 2017 at 15:05 GMT_

----

I think this is a little more complicated than I had hoped it would be. What if we get rid of the hybrid module concept and instead swap out sys.\_\_dict\_\_ something like this:

```
# lib/sys.py
from __go__.grumpy import SysDict, SysModules
SysModules['sys'].__dict__ = SysDict
... carry on with the module as normal
```

Currently \_\_dict\_\_ is not settable, but it will be once https://github.com/google/grumpy/pull/331 is merged.

**Comment by alanjds** [comment]
<a href="https://github.com/alanjds"><img src="https://avatars3.githubusercontent.com/u/155623?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [alanjds](https://github.com/alanjds)**
_Monday Jun 19, 2017 at 18:46 GMT_

----

Hum... the whole point of this PR is allow ``sys.stdout = StringIO()`` to propagate to ``print``.

If ``SysModules['sys'].__dict__ = SysDict`` fixes this need and is simpler, lets do it!

Seems to still need an ``[SysDict[k] = locals()[k] for k in __all__]`` at the last line, but for me it looks more ok than the HybridModule stuff.

**Comment by alanjds** [comment]
<a href="https://github.com/trotterdylan"><img src="https://avatars2.githubusercontent.com/u/117627?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [trotterdylan](https://github.com/trotterdylan)**
_Monday Jun 19, 2017 at 19:58 GMT_

----

Right, so the trick would be get rid of the Stdout, Stdin and Stderr grumpy package members and instead pull those from SysDict like `SysDict.GetItemString(f, "stdin")`. I have a change started to see how this would work in practice. I can either finish that up and create a PR, or if you like, you can go that route in this PR.

> Seems to still need an [SysDict[k] = locals()[k] for k in __all__] at the last line, but for me it looks more ok than the HybridModule stuff.

I don't think this is necessary if we do the \_\_dict\_\_ assignment at the very top of the module.
